ompi/mca/coll/han/coll_han_dynamic_file.h

```diff
@@ -0,0 +1,111 @@
+ * topological level. The standard name for these parameters is:
+ * [collective]_dynamic_[topologic_level]_module
+ *
+ * #######################
+ * # Dynamic file format #
+ * #######################
+ * File defined rules precede MCA parameter defined rule.
+ * To activate file reader, the MCA parameter use_dynamic_file_rules must
+ * be set to true. The path to the dynamic file is given by the MCA
+ * parameter dynamic_rules_filename. If there is any issue reading the file,
+ * the file is considered as invalid and only MCA parameter defined rules are
+ * used. If a potential logical issue is identified in the file, a
+ * warning is printed but the file is not considered as invalid.
+ *
+ * The file is built recursively.
+ * A set of rules of a type is built as follows:
+ * Number of rules of the set
+ * Rule1
+ * Rule2
+ * ...
+ *
+ * A rule of the level i is built as follows (excluding message size rule):
+ * Rule property
+ * Set of rules of level i+1
+ *
+ * A message size rule is built as follows:
+ * Message_size Component
+ *
+ * Rule properties are (by increasing level):
+ *     - Collective identifier:
+ *           Defined in ompi/mca/coll/base/coll_base_functions.h.
+ *     - Topologic level:
+ *           Defined in coll_han_dynamic.h. It defines the communicator
+ *           topology level. This is GLOBAL_COMMUNICATOR for the user
+ *           communicator and the corresponding level for sub-communicators
+ *           created by han.
+ *     - Configuration size:
+ *           The configuration size is the number of elements of the actual
+ *           topology level in the upper topology level. For example, if
+ *           topology levels are intra-node and inter-node, it can be the
+ *           number of MPI ranks per node or the number of nodes in the global
+ *           communicator. For the GLOBAL_COMMUNICATOR topologic level,
+ *           the configuration size is the communicator size.
+ *     - Message_size Component:
+ *           This is the message size, in bytes, of the message. Component is
+ *           the component identifier to use for this collective on this
+ *           communicator with this message size. Components identifier are
+ *           defined in coll_han_dynamic.h
+ *
+ * Here is an example of a dynamic rules file:
+ * 2 # Collective count
+ * 7 # Collective identifier 1 (defined in ompi/mca/coll/base/coll_base_functions.h)
+ * 2   # Topologic level count
+ * 0   # Topologic level identifier 1
+ * 1     # Configuration count
+ * 1     # Configuration size 1
+ * 2       # Message size rules count
+ * 0 3     # Message size 1 and component identifier
+ * 128 1   # Message size 2 and component identifier
+ * 1   # Topologic level identifier 2
+ * 1     # Configuration count
+ * 1     # Configuration size 1
+ * 1       # Message size rules count
+ * 0 1     # Message size 1 and component identifier
+ * 3 # Collective identifier 2
```

Comment:
It would be indeed a nice extension, but I don't thing it would be possible to implement this before May 14th: what if the component/the collective/the algorithm does not support segmentation ? 
Because of coll tuned configuration file asks for similar, I am afraid of falling in similar traps ("faninout" complexifies all coll/tuned tunings and seems to be almost never used).
I would prefer to let algorithms that handle segment size to provide a MCA threshold parameters.
The idea of this file is to decide "when" to call a algorithm but not "how". The assumption is made that the "how" is handled inside the algorithms : segmented algorithms may ask the topology hints for the segments size.

Reply:
I think there are two-level of segmentations here.

The first level is the segmentation in the HAN module.
Suppose we want to do a two-level hierarchical broadcast with 4MB message.
If we do not have the segmentation in the HAN module, this broadcast will become a 4MB inter-node broadcast followed by a 4MB intra-node broadcast (same as the algorithm implemented in mca_coll_han_bcast_intra_simple). In this way, there is no overlap between the inter-node and intra-node broadcast. For small messages, this is OK; but for larger messages, we can improve the performance by segmentation.
If we divide the 4MB message into 4 segments, 1MB each, a pipeline is formed as in mca_coll_han_bcast_intra. 
(ib is a nonblocking internode broadcast and sb is an intra-node broadcast)
iter 0 |  ib(seg0) |
iter 1 |  sb(seg0) | ib(seg1) |
iter 2 |                | sb(seg1) | ib(seg2) |
iter 3 |                |                | sb(seg2)| ib(seg3) |
iter 4 |                |                |              | sb(seg3) |
And if we choose the modules on the sub-communicators correctly, the inter-node and intra-node broadcast on those 1MB segments would overlap with each other and the overall performance of the 4MB broadcast would be improved.
This level of segmentation belongs to the HAN module, and that's why I think segment size is needed in this file.

The second level is the segmentation of segments in the HAN module,  the segmentation of those 1MB segments.
For example, when performing the 1MB inter-node Ibcast, like ib(seg1), what is the best segment size for it? 
For this kind of segmentation, I totally agree with your strategy. 
Originally, I handled this kind of segmentation in the HAN module as well (line 99 - 107 in mca_coll_han_bcast_intra).
But I think letting the modules and the algorithms handle "how" is a better way.

Reply:
I agree that pipelining segments is very interesting, but I am not convinced that the configuration file needs the segment size.
In your example of a 4MB bcast, if you want to use a segmented algorithm from Han to monitor each level, you set in the file the "global" topological-level of bcast to Han for 4MB. Not activating "simple" version of bcast  will call mca_coll_han_bcast_intra. This function will manage all the calls to the designed bcast for inter and intra topological level (according to the file configuration). So if you find out the segments must have a different size for each level it can be done inside this function.
Other components that handle segmentation already have their inputs for segment size (coll tuned config file for example). A segment size in this file would just be useful when Han is asked. And asking for another component with a specific segment size in this file would introduce conflicts with the other parameters. So I would rather prefer a set of Han MCA parameters or a separate file to guide specifically Han segmentation.